luigi/contrib/lsf.py

```diff
@@ -0,0 +1,362 @@
+        - Runner function loads the class from pickle
+        - Runner class untars the dependencies
+        - Runner function hits the button on the class's work() method
+        """
+        self._init_local()
+        self._run_job()
+
+    def work(self):
+        """
+        Subclass this for where you're doing your actual work.
+
+        Why not run(), like other tasks? Because we need run to always be
+        something that the Worker can call, and that's the real logical place to
+        do LSF scheduling.
+        So, the work will happen in work().
+        """
+        pass
+
+    def _dump(self, out_dir=''):
+        """
+        Dump instance to file.
+        """
+        self.job_file = os.path.join(out_dir, 'job-instance.pickle')
+        if self.__module__ == '__main__':
+            dump_inst = pickle.dumps(self)
+            module_name = os.path.basename(sys.argv[0]).rsplit('.', 1)[0]
+            dump_inst = dump_inst.replace('(c__main__', "(c" + module_name)
+            open(self.job_file, "w").write(dump_inst)
+
+        else:
+            pickle.dump(self, open(self.job_file, "w"))
+
+    def _run_job(self):
+        """
+        Build a bsub argument that will run lsf_runner.py on the directory we've specified.
+        """
+
+        args = []
+
+        if isinstance(self.output(), list):
+            log_output = self.output()[0].path.split("/")
+        else:
+            log_output = self.output().path.split("/")
+
+        args += ["bsub", "-q", self.queue_flag]
+        args += ["-n", str(self.n_cpu_flag)]
+        args += ["-M", str(self.memory_flag)]
+        args += ["-R", "rusage[%s]" % self.resource_flag]
+        args += ["-W", str(self.runtime_flag)]
+        if self.job_name_flag:
+            args += ["-J", str(self.job_name_flag)]
+        args += ["-o", "/".join(log_output[0:-1]) + "/job.out"]
+        args += ["-e", "/".join(log_output[0:-1]) + "/job.err"]
+        args += self.extra_bsub_args.split()
+
+        # Find where our file is
+        runner_path = lsf_runner.__file__
+        # assume source is next to compiled
+        if runner_path.endswith("pyc"):
+            runner_path = runner_path[:-3] + "py"
+
+        args += [runner_path]
+        args += [self.tmp_dir]
+
+        # That should do it. Let the world know what we're doing.
+        LOGGER.info(" ".join([str(a) for a in args]))
+
+        # Submit the job
+        run_job_proc = subprocess.Popen(
+            [str(a) for a in args],
+            stdin=subprocess.PIPE, stdout=subprocess.PIPE, cwd=self.tmp_dir)
+        output = run_job_proc.communicate()[0]
+
+        # ASSUMPTION
+        # The result will be of the format
+        # Job <123> is submitted ot queue <myqueue>
+        # So get the number in those first brackets.
+        # I cannot think of a better workaround that leaves logic on the Task side of things.
+        LOGGER.info("### JOB SUBMISSION OUTPUT:" + str(output))
```

Comment:
Since you're going string concatenation, your output will not include a space between `:` and your output var.

General critique here... You using both string concatenation and `%` interpolation throughout. Pick one (preferably `%` or `.format` (I'm personally partial to format)) and stick with it please.

Reply:
Sorry, but I went for the `%` version as the majority of instances already used that form. Happy to change if it matches in with the rest of the code base though.